Branch-SDK/src/io/branch/referral/Branch.java

```diff
@@ -1885,111 +1939,228 @@ public void getShortUrl(String channel, String feature, String stage, JSONObject
      *                 the link that will be passed into a new app session when clicked
      * @param callback A {@link BranchLinkCreateListener} callback instance that will trigger
      *                 actions defined therein upon receipt of a response to a create link request.
+     * @param  async   A {@link Boolean} whose value is true if the link should be created asynchronously.
+     *
      * @see BranchLinkData
      * @see BranchLinkData#putTags(Collection)
      * @see BranchLinkData#putChannel(String)
      * @see BranchLinkData#putFeature(String)
      * @see BranchLinkData#putStage(String)
      * @see BranchLinkData#putParams(String)
+     * @see BranchLinkData#putDuration(int)
      * @see BranchLinkCreateListener
+     * @deprecated use {@link BranchShortLinkBuilder} instead.
      */
-    public void getShortUrl(Collection<String> tags, String channel, String feature, String stage, JSONObject params, BranchLinkCreateListener callback) {
-        generateShortLink(null, LINK_TYPE_UNLIMITED_USE, 0, tags, channel, feature, stage, stringifyParams(filterOutBadCharacters(params)), callback, true);
+    @Deprecated
+    private String generateShortLink(final String alias, final int type, final int duration, final Collection<String> tags, final String channel, final String feature, final String stage, final String params, BranchLinkCreateListener callback, boolean async) {
+        ServerRequestCreateUrl req = new ServerRequestCreateUrl(context_, alias, type, duration, tags,
+                channel, feature, stage,
+                params, callback, async);
+
+        if (!req.constructError_ && !req.handleErrors(context_)) {
+            if (linkCache_.containsKey(req.getLinkPost())) {
+                String url = linkCache_.get(req.getLinkPost());
+                if (callback != null) {
```

Comment:
I'm not sure what it being private has to do with it, I think this is a bad choice for us to be using internally as well :stuck_out_tongue: I think this is going to lead to bugs within the SDK, since we're mixing code flow control patterns. If you'd rather, this can be refactored later, but I think this is introducing technical debt.